cpp/include/cudf/utilities/span.hpp

```diff
@@ -108,20 +100,91 @@ class span_base {
   size_type _size;
 };
 
+// ===== host_span =================================================================================
+
+namespace {
+
+template <typename T>
+struct is_host_span_container : std::false_type {
+};
+
+template <typename T, typename Alloc>
+struct is_host_span_container<  //
+  std::vector<T, Alloc>> : std::true_type {
+};
+
+template <typename T, typename Alloc>
+struct is_host_span_container<  //
+  thrust::host_vector<T, Alloc>> : std::true_type {
+};
+
+}  // namespace
+
 template <typename T, std::size_t Extent = dynamic_extent>
 struct host_span : public span_base<T, Extent, host_span<T, Extent>> {
   using base = cudf::detail::span_base<T, Extent, host_span<T, Extent>>;
   using base::base;
 
   constexpr host_span() noexcept : base() {}
+
+  template <typename C, std::enable_if<is_host_span_container<C>::value>* = nullptr>
+  constexpr host_span(C& in) : base(in.data(), in.size())
+  {
+  }
+
+  template <typename C, std::enable_if<is_host_span_container<C>::value>* = nullptr>
+  constexpr host_span(C const& in) : base(in.data(), in.size())
+  {
+  }
 };
 
+// ===== device_span ===============================================================================
+
+namespace {
+
+template <typename T>
+struct is_device_span_container : std::false_type {
+};
+
+template <typename T, typename Alloc>
+struct is_device_span_container<  //
+  thrust::device_vector<T, Alloc>> : std::true_type {
+};
+
+template <typename T>
+struct is_device_span_container<  //
+  rmm::device_vector<T>> : std::true_type {
+};
+
+template <>
+struct is_device_span_container<  //
+  rmm::device_buffer> : std::true_type {
+};
+
+}  // namespace
+
 template <typename T, std::size_t Extent = dynamic_extent>
 struct device_span : public span_base<T, Extent, device_span<T, Extent>> {
   using base = cudf::detail::span_base<T, Extent, device_span<T, Extent>>;
   using base::base;
 
   constexpr device_span() noexcept : base() {}
+
+  template <typename C, std::enable_if<is_device_span_container<C>::value>* = nullptr>
+  constexpr device_span(C& in)
+    : base(static_cast<T*>(  //
```

Comment:
Why is the cast needed?

Reply:
`rmm::device_buffer`

Reply:
`std::byte` is defined as `unsigned char` in c++ 17.
We could just _not support_ `rmm::device_buffer`, but we'd be passing a `void*` and `size_t` explicitly.

Reply:
My initial reaction is that `device_buffer` should not be supported via an implicit construction. It feels like it should be the user's responsibility to add the type information to `device_buffer.data()` first. 

Otherwise constructing from `device_buffer` would do more than "one thing": it imbues a type and views the data. All the other container ctors have their type imbued.

Reply:
I was torn between helpfulness and type safety on this one. I leaned toward helpfulness because `device_buffer` is known to provide `void*` for `data()`. Essentially we end up with either:
```
call_with_span(detail::device_span<unsigned char const>(b));
// or
call_without_span(b.size(), static_cast<unsigned char const*>(b.data()));
```
And both must explicitly name "unsigned char const". I'm not sure how helpful it is, though. We may just _want_ `void*`, and we can't support that with `span` unless we break compat with `std::span` (which we are sort of doing with device_span anyway...).

Reply:
Happy to remove this from `device_span`, since the usefullness of `void*` => `unsigned char*` conversion is questionable.

Reply:
I'd say let's disable support for `device_buffer` for now. There shouldn't be many cases where a `device_buffer` is being directly passed into a place where a `device_span` is expected. It's more likely that a `device_vector` or `device_uvector` will be used.

This also reminds me we should add an implicit conversion operator from `column_view` to `device_span` (though it needs to check to ensure the data is fixed width and non-null, otherwise throws an exception). This can be a follow up PR.

Reply:
Removed, and added `device_uvector` since it was missing.